cpp/dolfinx/io/checkpointing.cpp

```diff
@@ -0,0 +1,354 @@
+    {"unset", basix::element::lagrange_variant::unset},
+    {"equispaced", basix::element::lagrange_variant::equispaced},
+    {"gll_warped", basix::element::lagrange_variant::gll_warped},
+};
+
+} // namespace
+
+namespace dolfinx::io::checkpointing
+{
+//-----------------------------------------------------------------------------
+template <std::floating_point T>
+void write_mesh(ADIOS2Wrapper& ADIOS2, dolfinx::mesh::Mesh<T>& mesh)
+{
+
+  const mesh::Geometry<T>& geometry = mesh.geometry();
+  std::shared_ptr<const mesh::Topology> topology = mesh.topology();
+
+  // Variables/attributes to save
+  std::int32_t dim = geometry.dim();
+  std::uint64_t num_vertices_global, offset, num_cells_global, cell_offset;
+  std::uint32_t num_vertices_local, num_cells_local, num_dofs_per_cell, degree;
+  std::string cell_type, lagrange_variant;
+  std::vector<std::int64_t> array_global;
+  std::vector<std::int32_t> offsets_global;
+
+  std::shared_ptr<const common::IndexMap> geom_imap;
+
+  // Vertices information
+  {
+    geom_imap = mesh.geometry().index_map();
+    num_vertices_global = geom_imap->size_global();
+    num_vertices_local = geom_imap->size_local();
+    offset = geom_imap->local_range()[0];
+  }
+
+  // Cells information
+  {
+    const std::shared_ptr<const common::IndexMap> topo_imap
+        = topology->index_map(dim);
+    num_cells_global = topo_imap->size_global();
+    num_cells_local = topo_imap->size_local();
+    cell_offset = topo_imap->local_range()[0];
+  }
+
+  // Coordinate element information
+  {
+    const fem::CoordinateElement<T>& cmap = mesh.geometry().cmap();
+    fem::ElementDofLayout geom_layout = cmap.create_dof_layout();
+    num_dofs_per_cell = geom_layout.num_entity_closure_dofs(dim);
+    cell_type = mesh::to_string(cmap.cell_shape());
+    degree = cmap.degree();
+    lagrange_variant = variant_to_string[cmap.variant()];
+  }
+
+  const std::vector<int64_t> input_global_indices
+      = geometry.input_global_indices();
+  const std::span<const int64_t> input_global_indices_span(
+      input_global_indices.begin(), num_vertices_local);
+  const std::span<const T> mesh_x = geometry.x();
+
+  // Connectivity
+  {
+    std::shared_ptr<const graph::AdjacencyList<std::int32_t>> connectivity
+        = topology->connectivity(dim, 0);
+    const std::vector<std::int32_t>& array = connectivity->array();
+    const std::vector<std::int32_t>& offsets = connectivity->offsets();
+
+    const std::span<const int32_t> array_span(array.begin(),
+                                              offsets[num_cells_local]);
+
+    array_global.resize(offsets[num_cells_local]);
+    offsets_global.resize(num_cells_local + 1);
+
+    std::iota(array_global.begin(), array_global.end(), 0);
+
+    geom_imap->local_to_global(array_span, array_global);
+
+    for (std::size_t i = 0; i < num_cells_local + 1; ++i)
+      offsets_global[i] = offsets[i] + cell_offset * num_dofs_per_cell;
+  }
```

Comment:
I do not think the connectivity here is quite right.
You would have to map between the `mesh->topology()` and `mesh->geometry()` indices to make this correct.
This is for instance what is done in `dolfinx::mesh::entities_to_geometry`: https://github.com/FEniCS/dolfinx/blob/8e4122b5b28a8997b94148797125becf5ba704cb/cpp/dolfinx/mesh/utils.h#L634
which  you in turn can use `geom_imap->local_to_global` on. 

I don't even think we need this, as you could simply use the `geometry.dofmap()`  (https://github.com/FEniCS/dolfinx/blob/8e4122b5b28a8997b94148797125becf5ba704cb/cpp/dolfinx/mesh/Geometry.h#L121-L134) to get the local indices per cell and then in turn call `geom_imap->local_to_global(...)` on the array created from this mdspan.

Reply:
Thanks. Updated to using `geometry.dofmap()`.